### DISCUSSION TITLE
<How to use React Router Dom with a project?>

### DISCUSSION THREAD

**Discussion by RafeedMohammad11** [community/community#123587] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

How to use React Router Dom with a project?

**Comment by AsibHasanRiyad** [comment]
### Step 1: Install `react-router-dom`

Make sure you have `react-router-dom` installed. You can install it using npm or yarn:

```bash
npm install react-router-dom
# or
yarn add react-router-dom
```

### Step 2: Set Up Routes

In your main application file (e.g., `App.js`), set up your routes using `BrowserRouter`, `Route`, and `Switch` from `react-router-dom`.

```jsx
// Import necessary modules
import React from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';

// Import your components
import Home from './components/Home';
import About from './components/About';
import Contact from './components/Contact';
import NotFound from './components/NotFound';

// Define your App component
function App() {
  return (
    <BrowserRouter>
      <div className="App">
        {/* Your navigation or header can go here */}

        {/* Define your routes */}
        <Switch>
          {/* Route for Home component */}
          <Route exact path="/" component={Home} />

          {/* Route for About component */}
          <Route path="/about" component={About} />

          {/* Route for Contact component */}
          <Route path="/contact" component={Contact} />

          {/* Route for handling 404 (Not Found) */}
          <Route component={NotFound} />
        </Switch>
      </div>
    </BrowserRouter>
  );
}

export default App;
```

### Step 3: Create Components

Create the components (`Home.js`, `About.js`, `Contact.js`, `NotFound.js`) in a `components` directory within your project.

**Example: Home.js**

```jsx
import React from 'react';

const Home = () => {
  return (
    <div>
      <h2>Welcome to the Home Page</h2>
      {/* Your home page content goes here */}
    </div>
  );
}

export default Home;
```

**Example: About.js**

```jsx
import React from 'react';

const About = () => {
  return (
    <div>
      <h2>About Us</h2>
      {/* Your about page content goes here */}
    </div>
  );
}

export default About;
```

**Example: Contact.js**

```jsx
import React from 'react';

const Contact = () => {
  return (
    <div>
      <h2>Contact Us</h2>
      {/* Your contact page content goes here */}
    </div>
  );
}

export default Contact;
```

**Example: NotFound.js**

```jsx
import React from 'react';

const NotFound = () => {
  return (
    <div>
      <h2>404 - Not Found</h2>
      {/* Display a message or a link to go back to home */}
    </div>
  );
}

export default NotFound;
```

### Step 4: Link to Routes

In your application, use `Link` components from `react-router-dom` to navigate between routes.

**Example: Navigation Component**

```jsx
import React from 'react';
import { Link } from 'react-router-dom';

const Navigation = () => {
  return (
    <nav>
      <ul>
        <li><Link to="/">Home</Link></li>
        <li><Link to="/about">About</Link></li>
        <li><Link to="/contact">Contact</Link></li>
      </ul>
    </nav>
  );
}

export default Navigation;
```

### Step 5: Start Your Application

Run your React application using `npm start` or `yarn start` and navigate to different routes to see how React Router DOM manages the routing.

That's it! This setup should give you a basic understanding of how to use React Router DOM in your project. Feel free to expand upon this example by adding more routes, nested routes, or using other features provided by `react-router-dom` based on your project's requirements.

**Comment by Pulkit1822** [comment]
**Using React Router Dom in Your Project: A Beginner's Guide**
============================================================

Hey there! Today, I'll guide you through using React Router Dom in your project. Let's get started!

### Step 1: Install React Router Dom

First things first, you need to install the `react-router-dom` package in your project. You can do this using npm or yarn:

```bash
npm install react-router-dom
```

or

```bash
yarn add react-router-dom
```

### Step 2: Set up the Router

In your main `index.js` file, import the `BrowserRouter` component from `react-router-dom` and wrap your root component with it:

```jsx
import { BrowserRouter } from 'react-router-dom';

ReactDOM.render(
  <BrowserRouter>
    <App />
  </BrowserRouter>,
  document.getElementById('root')
);
```

This sets up the routing features for your application.

### Step 3: Define Routes

In your `App.js` file (or any component where you want to define routes), import the `Routes` and `Route` components from `react-router-dom`. Then, define your routes inside the `Routes` component:

```jsx
import { Routes, Route } from 'react-router-dom';
import Home from './components/Home';
import About from './components/About';
import Contact from './components/Contact';

function App() {
  return (
    <div>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/about" element={<About />} />
        <Route path="/contact" element={<Contact />} />
      </Routes>
    </div>
  );
}
```

Here, we've defined three routes: `/`, `/about`, and `/contact`, each rendering a different component.

### Step 4: Use Link to Navigate

To create links between your routes, use the `Link` component from `react-router-dom`:

```jsx
import { Link } from 'react-router-dom';

function Navbar() {
  return (
    <nav>
      <ul>
        <li>
          <Link to="/">Home</Link>
        </li>
        <li>
          <Link to="/about">About</Link>
        </li>
        <li>
          <Link to="/contact">Contact</Link>
        </li>
      </ul>
    </nav>
  );
}
```

The `to` prop in `Link` specifies the path to navigate to when the link is clicked.

### Step 5: Access Route Parameters

If you need to access parameters in your route paths, use the `:paramName` syntax in the path and access the parameter using the `useParams` hook:

```jsx
import { Routes, Route, useParams } from 'react-router-dom';

function UserProfile() {
  const { userId } = useParams();

  return <div>User Profile for user {userId}</div>;
}

function App() {
  return (
    <Routes>
      <Route path="/users/:userId" element={<UserProfile />} />
    </Routes>
  );
}
```

In this example, the route path `/users/:userId` defines a parameter named `userId`. The `useParams` hook allows you to access the value of `userId` in the `UserProfile` component.

That's it! You're now equipped to use React Router Dom in your project. Feel free to explore more advanced features like nested routes and redirects to enhance your routing capabilities.

**Mark as Correct Answer**
-------------------------

If you found this guide helpful, please mark me as the correct answer. Happy coding!
